core/amber/src/main/scala/edu/uci/ics/texera/workflow/common/storage/memory/JCSOpResultStorage.scala

```diff
@@ -0,0 +1,62 @@
+package edu.uci.ics.texera.workflow.common.storage.memory
+
+import com.typesafe.scalalogging.Logger
+import edu.uci.ics.texera.workflow.common.storage.OpResultStorage
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+import org.apache.commons.jcs3.JCS
+import org.apache.commons.jcs3.access.CacheAccess
+import sun.reflect.generics.reflectiveObjects.NotImplementedException
+
+import java.util.concurrent.locks.ReentrantLock
+
+class JCSOpResultStorage extends OpResultStorage {
+
+  private val logger = Logger(this.getClass.getName)
+
+  private val lock = new ReentrantLock()
+
+  private val cache: CacheAccess[String, List[Tuple]] = JCS.getInstance("texera")
+
+  override def put(key: String, records: List[Tuple]): Unit = {
+    lock.lock()
```

Comment:
For MongoDB implementation, add try catch to release lock.
For JCS and In-Memory, try to see if this can be changed using concurrent data structures.

Reply:
In ConcurrentHashMap, according to [this link](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/concurrent/ConcurrentHashMap.html)
> Retrieval operations (including get) generally do not block, so may overlap with update operations (including put and remove)
and in TireMap, read operation is not guaranteed atomic. Currently, I'm using ConcurrentHashMap.